Anthony's changes - making a PR req will allow for simpler comparison

AreezG opened the pull request:
making a PR req will allow for simpler comparison between old and new files

anthonymccanny committed:
Initial Upload - data_cleaner and tweaks

data_cleaner, the merged cleaning file is uploaded along with placeholder for the main code running file and some tweaks to table5 and table7

anthonymccanny committed:
Master file created

Master file created and other files edited to match the change. All code should be runnable by running main.do. All code uses data_cleaner.do for cleaning

AreezG commented on data_cleaner.do:
```diff
@@ -0,0 +1,285 @@
+gen final_city = "."
+replace final_city = clean_hcity if good_city==1
+
+// ANTHONY EDIT - THIS CODE APPEARS TO DO NOTHING
```
I think this was an attempt to do some zip matching but we didn't go further with it. We should most likely delete it.

AreezG commented on data_cleaner.do:
```diff
@@ -0,0 +1,285 @@
+	// Generate uppercase values of hcity and primary_city
+	// These will be used later to match on subsections of the string
+	gen upper_hcity = upper(hcity)
+    gen upper_primary = upper(primary_city)
```
tab over for consistency

AreezG commented on data_cleaner.do:
```diff
@@ -0,0 +1,285 @@
+        gen `newvarname' = `varname'
+		
+		// Loop over a set of characters
+		foreach char in " " "." "-" "'" "," "0" "1" "2" "3" "4" "5" "6" "7" "8" "9" {
```
can we do this in a more regexy way? since we're only keeping letters maybe we can do an inrange to be more specific: https://www.statalist.org/forums/forum/general-stata-discussion/general/3923-drop-special-characters-from-a-string-variable

mattdwebb replied:
Given the comments below, I think we need to deal with the zipcode hcity issue first.  That is, take out all the items where people coded a zip code as a city.  Then we can do something like the above, then do the remaining manual changes, removing the special characters.

AreezG replied:
i think that makes sense for the zips. do those first. 

i still lean more towards selecting which chars to include vs exclude though, it'll be more robust for what we want to do (what if there are `/` or `\` or `:` which are all characters that might honestly be in the dataset or in future versions of the dataset? it'll be more replicable and reusable to say which chars to include)

AreezG commented on data_cleaner.do:
```diff
@@ -0,0 +1,285 @@
+    clean_city officialuspscityname clean_usps // Clean USPS city name
+
+    replace good_city = 1 if clean_hcity == clean_primary // Flag city name as good if it matches primary city
+	replace good_city = 1 if clean_hcity == clean_usps // Flag city name as good if it matches USPS city name
```
spacing

AreezG commented on data_cleaner.do:
```diff
@@ -0,0 +1,285 @@
+    replace good_city = 1 if clean_hcity == clean_primary // Flag city name as good if it matches primary city
+	replace good_city = 1 if clean_hcity == clean_usps // Flag city name as good if it matches USPS city name
+
+	// Loop over all the acceptable cities
+	forvalues i = 1 /16 {
+		clean_city allaccept`i' clean_allaccept`i'
+		replace good_city = 1 if clean_hcity == clean_allaccept`i'
+	}
+
+	// Loop over all the acceptable counties
+	forvalues i = 1 /6 {
+		clean_city allcounty`i' clean_allcounty`i'
+		replace good_city = 1 if clean_hcity == clean_allcounty`i'
+	}
```
again spacing

AreezG commented on data_cleaner.do:
```diff
@@ -0,0 +1,285 @@
+// Create the final city
+gen final_city = "."
+replace final_city = clean_hcity if good_city==1
+
```
now we're too snug!

AreezG commented on data_cleaner.do:
```diff
@@ -0,0 +1,285 @@
+	"MTOLIVE>MOUNTOLIVE " ///
+	"MTOLIVETWP>MOUNTOLIVE " ///
+	"PARSIPPANY-TROYHILLS>PARSIPPANY " ///
+	"BAYRIDGE,BROOKLYN>BROOKLYN " ///
```
we removed the commas right?  should remove here too

AreezG commented on data_cleaner.do:
```diff
@@ -0,0 +1,285 @@
+	"SANANTONIO,TX>SANANTONIO " ///
+	"DELMARHEIGHTS>DELMAR " ///
+	"SOUTHSANJOSE>SANJOSE " ///
+	"33614>TAMPA " ///
```
how does this work if we removed all numbers? should we keep numbers above? sorry i am bad at reading stata code

AreezG commented on data_cleaner.do:
```diff
@@ -0,0 +1,285 @@
+	"33614>TAMPA " ///
+	"33511>BRANDON " ///
+	"CLEARWATERBEACH>CLEARWATER " ///
+	"PALM/HUDSON>PALMHARBOR " ///
```
noting more punctuation, these comments might be my misunderstanding though

AreezG commented on data_cleaner.do:
```diff
@@ -0,0 +1,285 @@
+	"DELMARHEIGHTS>DELMAR " ///
+	"SOUTHSANJOSE>SANJOSE " ///
+	"33614>TAMPA " ///
+	"33511>BRANDON " ///
```
numbers

AreezG commented on data_cleaner.do:
```diff
@@ -0,0 +1,285 @@
+	"WARRINGTONLANE>WARRINGTON " ///
+	"WARRINGTONTWP>WARRINGTON " ///
+	"CHESTEFILED>CHESTERFIELD " ///
+	"92284>YUCCAVALLEY " ///
```
numbers

AreezG commented on data_cleaner.do:
```diff
@@ -0,0 +1,285 @@
+	"WARRINGTONTWP>WARRINGTON " ///
+	"CHESTEFILED>CHESTERFIELD " ///
+	"92284>YUCCAVALLEY " ///
+	"SANANTONIO,TX>SANANTONIO " ///
```
punctuation

AreezG commented on data_cleaner.do:
```diff
@@ -0,0 +1,285 @@
+	"CLEARWATERBEACH>CLEARWATER " ///
+	"PALM/HUDSON>PALMHARBOR " ///
+	"REDINGTONSHORES>REDINGTONBEACH " ///
+	"BEACHWOODOH44122>BEACHWOOD " ///
```
numbers

AreezG commented on data_cleaner.do:
```diff
@@ -0,0 +1,285 @@
+	// Define a new program 'clean_city'
+	program define clean_city
+		args varname newvarname
+        gen `newvarname' = `varname'
```
tab over

AreezG commented on data_cleaner.do:
```diff
@@ -0,0 +1,285 @@
+replace final_city = clean_hcity if good_city==1
+
+// ANTHONY EDIT - THIS CODE APPEARS TO DO NOTHING
+/* // Try some more string groups using zip codes	
+gen strzip = zip
+tostring strzip, replace force
+gen ziplength = strlen(strzip)
+replace strzip = "0" + strzip if ziplength==4
+gen zipfour = substr(strzip,1,4) */
```
we should delete if we're not using

AreezG reviewed (commented):
Way cleaner than what we originally had. Thanks for cleaning this up. 

Most of what i'm seeing is just spacing and then some questions about the manual changes. there's one area where we're stripping certain characters but it might make more sense to include only the characters we want instead. there could be lots of errant punctuation in these kinds of datasets which we haven't noticed so i think a very strict only capital letters (and maybe numbers?) might be the way to go

anthonymccanny committed:
Cleaning Debug + hcity Control Fix

anthonymccanny committed:
Edits to main.do

AreezG commented on data_cleaner.do:
```diff
@@ -0,0 +1,274 @@
+        gen `newvarname' = `varname'
+		
+		// Loop over a set of characters
+		foreach char in " " "." "-" "'" "," "0" "1" "2" "3" "4" "5" "6" "7" "8" "9" "\\" {
```
might be unnecessary ( i imagine the effects would be limited) but do we want to check for a few more characters? e.g. 
`"`, `:` `#`, `$` `%`, `[`, `]`, `(`, `)`.

anthonymccanny replied:
Made the change! I don't think it changed anything, but it doesn't hurt to be complete!

AreezG commented on data_cleaner.do:
```diff
@@ -0,0 +1,274 @@
+	"PARSIPPANYTROYHILLS>PARSIPPANY " ///
+	"BAYRIDGEBROOKLYN>BROOKLYN " ///
+	"BKLYN>BROOKLYN " ///
+	"FLUSHINGQUEENS>FLUSHING " ///
```
I think that `BAYRIDGEBROOKLYN` , `BKLYN`, and `FLUSHINGQUEENS` should all be set to whatever New York City is. Brooklyn and Queens are borroughs, NYC would be more consistent.

anthonymccanny replied:
This is a great point, and in making the change I realized that we actually have a lot of the boroughs as final city names in the current version of our cleaning code. Because each borough is so big, and the housing markets I assume are different, and we had good data on each of them, I though it could be reasonable to use the boroughs as the final unit of analysis. But the issue was what to do with entries that had just been marked as "New York" for city. So I did some digging and manual inspection and found that all entries that I checked (about 20) that were ending up as NEWYORK in the final version of the cleaning code were actually houses in Manhattan (which is colloquially sometimes just called New York City). So I decided to mark all New York houses with their boroughs for consistency, changing NEWYORK to MANHATTAN in all instances.

AreezG reviewed (commented):
a couple more comments here, otherwise LGTM! 🚢 :shipit:  🍏

anthonymccanny committed:
Final Commit Before Merge

Make sure that all NYC boroughs receive the name of the borough as the final city name; remove more special characters; fix output locations for table10.
